src/ru/nsu/ccfit/zuev/osu/game/Slider.java

```diff
@@ -283,31 +301,94 @@ public void init(final GameObjectListener listener, final Scene scene,
                 }
             }
 
-            abstractSliderBody.applyToScene(scene, Config.isSnakingInSliders());
-            abstractSliderBody.setBodyColor(bodyColor.r(), bodyColor.g(), bodyColor.b());
-            abstractSliderBody.setBorderColor(borderColor.r(), borderColor.g(), borderColor.b());
+            sliderBody.applyToScene(scene, Config.isSnakingInSliders());
+            sliderBody.setBodyColor(bodyColor.r(), bodyColor.g(), bodyColor.b());
+            sliderBody.setBorderColor(borderColor.r(), borderColor.g(), borderColor.b());
+        }
+
+        if (Config.isDimHitObjects()) {
+            isCircleDimming = true;
+
+            // Source: https://github.com/peppy/osu/blob/60271fb0f7e091afb754455f93180094c63fc3fb/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs#L101
+            var dimDelaySec = ((float) beatmapSlider.timePreempt / 1000f - objectHittableRange) / GameHelper.getSpeedMultiplier();
+            var colorDim = 195f / 255f;
+
+            applyCircleDimming(startCircle, startOverlay, dimDelaySec, comboColor);
+            applyCircleDimming(endCircle, endOverlay, dimDelaySec, comboColor);
+
+            number.setColor(colorDim, colorDim, colorDim);
+            number.registerEntityModifier(Modifiers.sequence(
+                Modifiers.delay(dimDelaySec),
+                Modifiers.color(0.1f / GameHelper.getSpeedMultiplier(),
+                    number.getRed(), 1f,
+                    number.getGreen(), 1f,
+                    number.getBlue(), 1f
+                )
+            ));
+
+            endArrow.setColor(colorDim, colorDim, colorDim);
+            endArrow.registerEntityModifier(Modifiers.sequence(
+                Modifiers.delay(dimDelaySec),
+                Modifiers.color(0.1f / GameHelper.getSpeedMultiplier(),
+                    endArrow.getRed(), 1f,
+                    endArrow.getGreen(), 1f,
+                    endArrow.getBlue(), 1f
+                )
+            ));
+
+            sliderBody.applyDimAnimations(dimDelaySec);
+        } else {
+            isCircleDimming = false;
         }
 
         applyBodyFadeAdjustments(fadeInDuration);
     }
 
+    private void applyCircleDimming(Sprite circle, Sprite overlay, float dimDelaySec, RGBColor comboColor) {
+
+        var colorDim = 195f / 255f;
+
+        // Circle requires special handling because it's tinted with combo color.
+        circle.setColor(comboColor.r() * colorDim, comboColor.g() * colorDim, comboColor.b() * colorDim);
+        circle.registerEntityModifier(Modifiers.sequence(
+            Modifiers.delay(dimDelaySec),
+            Modifiers.color(0.1f / GameHelper.getSpeedMultiplier(),
+                circle.getRed(), comboColor.r(),
+                circle.getGreen(), comboColor.g(),
+                circle.getBlue(), comboColor.b()
+            )
+        ).setOnFinished(entity -> isCircleDimming = true));
```

Comment:
Same for here.

Reply:
Solved with [a77c024](https://github.com/osudroid/osu-droid/pull/401/commits/a77c02440c45372a4965d3691a79bb0a67be3ee8)